manager/allocator/networkallocator/networkallocator.go

```diff
@@ -376,593 +28,81 @@ func OnInit(options *ServiceAllocationOpts) {
 	options.OnInit = true
 }
 
-// ServiceNeedsAllocation returns true if the passed service needs to have network resources allocated/updated.
-func (na *NetworkAllocator) ServiceNeedsAllocation(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool {
-	var options ServiceAllocationOpts
-	for _, flag := range flags {
-		flag(&options)
-	}
+// NetworkAllocator provides network model specific allocation functionality.
+type NetworkAllocator interface {
+	//
+	// Drivers
+	//
 
-	specNetworks := serviceNetworks(s)
+	// IsBuiltInDriver returns whether the passed driver is an internal network driver
+	IsBuiltInDriver(name string) bool
 
-	// If endpoint mode is VIP and allocator does not have the
-	// service in VIP allocated set then it needs to be allocated.
-	if len(specNetworks) != 0 &&
-		(s.Spec.Endpoint == nil ||
-			s.Spec.Endpoint.Mode == api.ResolutionModeVirtualIP) {
+	//
+	// Network Allocation
+	//
 
-		if _, ok := na.services[s.ID]; !ok {
-			return true
-		}
+	// IsAllocated returns if the passed network has been allocated or not.
+	IsAllocated(n *api.Network) bool
 
-		if s.Endpoint == nil || len(s.Endpoint.VirtualIPs) == 0 {
-			return true
-		}
+	// Allocate allocates all the necessary resources both general
+	// and driver-specific which may be specified in the NetworkSpec
+	Allocate(n *api.Network) error
 
-		// If the spec has networks which don't have a corresponding VIP,
-		// the service needs to be allocated.
-	networkLoop:
-		for _, net := range specNetworks {
-			for _, vip := range s.Endpoint.VirtualIPs {
-				if vip.NetworkID == net.Target {
-					continue networkLoop
-				}
-			}
-			return true
-		}
-	}
+	// Deallocate frees all the general and driver specific resources
+	// which were assigned to the passed network.
+	Deallocate(n *api.Network) error
 
-	// If the spec no longer has networks attached and has a vip allocated
-	// from previous spec the service needs to allocated.
-	if s.Endpoint != nil {
-	vipLoop:
-		for _, vip := range s.Endpoint.VirtualIPs {
-			if na.IsVIPOnIngressNetwork(vip) && IsIngressNetworkNeeded(s) {
-				continue vipLoop
-			}
-			for _, net := range specNetworks {
-				if vip.NetworkID == net.Target {
-					continue vipLoop
-				}
-			}
-			return true
-		}
-	}
+	//
+	// Service Allocation
+	//
 
-	// If the endpoint mode is DNSRR and allocator has the service
-	// in VIP allocated set then we return to be allocated to make
-	// sure the allocator triggers networkallocator to free up the
-	// resources if any.
-	if s.Spec.Endpoint != nil && s.Spec.Endpoint.Mode == api.ResolutionModeDNSRoundRobin {
-		if _, ok := na.services[s.ID]; ok {
-			return true
-		}
-	}
+	// ServiceNeedsAllocation returns true if the passed service
+	// needs to have network resources allocated/updated.
+	ServiceNeedsAllocation(s *api.Service, flags ...func(*ServiceAllocationOpts)) bool
```

Comment:
change == all three changes suggested (to be clear `ServiceNeedsAllocation` -> `IsServiceAllocated`, `ServiceAllocate` -> `AllocateService`, `ServiceDeallocate`-> `DeallocateService`).

Note also that `ServiceNeedsAllocation` -> `IsServiceAllocated` inverts the sense, so I'll need to be careful...